velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -319,4 +319,99 @@ struct DayOfWeekFunction : public InitSessionTimezone<T>,
+struct MonthsBetweenFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Timestamp>* /*timestamp1*/,
+      const arg_type<Timestamp>* /*timestamp2*/,
+      bool /*roundOff*/) {
+    timeZone_ = getTimeZoneFromConfig(config);
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getYear(const std::tm& time) {
+    return 1900 + time.tm_year;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getMonth(const std::tm& time) {
+    return 1 + time.tm_mon;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getDay(const std::tm& time) {
+    return time.tm_mday;
+  }
+
+  FOLLY_ALWAYS_INLINE std::tuple<int32_t, int32_t, int32_t, int32_t>
+  splitTimestamp(Timestamp timestamp) {
+    std::tuple<int32_t, int32_t, int32_t, int32_t> result;
+    int32_t year = getYear(getDateTime(timestamp, this->timeZone_));
+    int32_t monthInYear = getMonth(getDateTime(timestamp, this->timeZone_));
+    int32_t dayInMonth = getDay(getDateTime(timestamp, this->timeZone_));
+    int32_t daysToMonthEnd = util::getMaxDayOfMonth(year, monthInYear) -
+        getDay(getDateTime(timestamp, this->timeZone_));
+    result = std::make_tuple(year, monthInYear, dayInMonth, daysToMonthEnd);
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t getSecondsInDay(
+      Timestamp timestamp,
+      int32_t year,
+      int32_t month,
+      int32_t day) {
+    const date::time_zone* timeZone = this->timeZone_;
+    Timestamp ts =
+        util::fromDatetime(util::daysSinceEpochFromDate(year, month, day), 0);
+    if (timeZone != nullptr) {
+      ts.toGMT(*timeZone);
+    }
+    return (timestamp.toMicros() - ts.toMicros()) / kNanosecondsInMillisecond;
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      double& result,
+      const arg_type<Timestamp>& timestamp1,
+      const arg_type<Timestamp>& timestamp2,
+      bool roundOff) {
+    int32_t year1{0};
+    int32_t monthInYear1{0};
+    int32_t dayInMonth1{0};
+    int32_t daysToMonthEnd1{0};
+    int32_t year2{0};
+    int32_t monthInYear2{0};
+    int32_t dayInMonth2{0};
+    int32_t daysToMonthEnd2{0};
+
+    std::tie(year1, monthInYear1, dayInMonth1, daysToMonthEnd1) =
+        splitTimestamp(timestamp1);
+    std::tie(year2, monthInYear2, dayInMonth2, daysToMonthEnd2) =
+        splitTimestamp(timestamp2);
+
+    int32_t months1 = year1 * 12 + monthInYear1;
+    int32_t months2 = year2 * 12 + monthInYear2;
+
+    double monthDiff = (double)(months1 - months2);
+    if (dayInMonth1 == dayInMonth2 ||
+        ((daysToMonthEnd1 == 0) && (daysToMonthEnd2 == 0))) {
+      result = monthDiff;
+    } else {
+      int64_t secondsInDay1 =
+          getSecondsInDay(timestamp1, year1, monthInYear1, dayInMonth1);
+      int64_t secondsInDay2 =
+          getSecondsInDay(timestamp2, year2, monthInYear2, dayInMonth2);
+      int64_t secondsDiff = (dayInMonth1 - dayInMonth2) * kSecondsInDay +
+          secondsInDay1 - secondsInDay2;
+      int64_t secondsInMonth = 31 * kSecondsInDay;
```

Comment:
Do we need to consider a month not containing 31 days?

Reply:
@rui-mo This part is consistent with the logic of spark.
https://github.com/apache/spark/blob/5b609598503df603cbddd5e1adf8d2cb28a5f977/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L333